File: leetcode/php/0005-longest-palindromic-substring.php
<?php

/**
 * 5. Longest Palindromic Substring
 * https://leetcode.com/problems/longest-palindromic-substring/
 */

class Solution
{
    /**
     * @param String $s
     * @return String
     */
    function longestPalindrome($s)
    {
        $ans = "";
        $anslen = 0;
        $len = strlen($s);

        for ($i = 0; $i < $len; $i++) {
            // substr with len = 1, 3, 5, ...
            $j = max(0, ceil($anslen / 2)); // set start j, start len must be greater that ans len
            $maxj = min($i, $len - $i);
            while ($j <= $maxj) {
                $str = substr($s, $i - $j, 1 + 2 * $j);
                if ($str != strrev($str)) break;
                if (strlen($str) > $anslen) {
                    $ans = $str;
                    $anslen = strlen($str);
                }
                $j++;
            }
            // substr with len = 2, 4, 6, ...
            $j = max(0, ceil(($anslen - 1) / 2)); // set start j, start len must be greater that ans len
            $maxj = min($i, $len - $i - 1);
            while ($j <= $maxj) {
                $str = substr($s, $i - $j, 2 + 2 * $j);
                if ($str != strrev($str)) break;
                if (strlen($str) > $anslen) {
                    $ans = $str;
                    $anslen = strlen($str);
                }
                $j++;
            }
        }

        return $ans;
    }
}

File: swift/0005-longest-palindromic-substring.swift
/**
 * 5. Longest Palindromic Substring
 * https://leetcode.com/problems/longest-palindromic-substring/
 **/

class Solution {
    func longestPalindrome(_ s: String) -> String {
        let chs = Array(s).map { String($0) }
        var best = [0,0]

        // a, aba, abcba, ...
        for i in 0..<chs.count {
            for j in 0...min(i, chs.count-i-1) {
                if chs[i-j] != chs[i+j] {
                    break
                }
                if best[1]-best[0] < 2*j {
                    best = [i-j, i+j]
                }
            }
        }

        // aa, abba, abccba, ...
        for i in 1..<chs.count {
            for j in 0...min(i-1, chs.count-i-1) {
                if chs[i-1-j] != chs[i+j] {
                    break
                }
                if best[1]-best[0] < 2*j+1 {
                    best = [i-1-j, i+j]
                }
            }
        }

        return Array(chs[best[0]...best[1]]).joined(separator: "")
    }
}